Firestore/core/src/firebase/firestore/local/leveldb_mutation_queue.mm

```diff
@@ -0,0 +1,519 @@
+    std::string index_prefix =
+        LevelDbDocumentMutationKey::KeyPrefix(user_id_, document_key.path());
+    for (index_iterator->Seek(index_prefix); index_iterator->Valid();
+         index_iterator->Next()) {
+      // Only consider rows matching exactly the specific key of interest. Index
+      // rows have this form (with markers in brackets):
+      //
+      // <User>user <Path>collection <Path>doc <BatchId>2 <Terminator>
+      // <User>user <Path>collection <Path>doc <BatchId>3 <Terminator>
+      // <User>user <Path>collection <Path>doc <Path>sub <Path>doc <BatchId>3
+      // <Terminator>
+      //
+      // Note that Path markers sort after BatchId markers so this means that
+      // when searching for collection/doc, all the entries for it will be
+      // contiguous in the table, allowing a break after any mismatch.
+      if (!absl::StartsWith(index_iterator->key(), index_prefix) ||
+          !row_key.Decode(index_iterator->key()) ||
+          row_key.document_key() != document_key) {
+        break;
+      }
+
+      batch_ids.insert(row_key.batch_id());
+    }
+  }
+
+  return AllMutationBatchesWithIds(batch_ids);
+}
+
+std::vector<FSTMutationBatch*>
+LevelDbMutationQueue::AllMutationBatchesAffectingDocumentKey(
+    const DocumentKey& key) {
+  // Scan the document-mutation index starting with a prefix starting with the
+  // given documentKey.
+  std::string index_prefix =
+      LevelDbDocumentMutationKey::KeyPrefix(user_id_, key.path());
+  auto index_iterator = db_.currentTransaction->NewIterator();
+  index_iterator->Seek(index_prefix);
+
+  // Simultaneously scan the mutation queue. This works because each (key,
+  // batchID) pair is unique and ordered, so when scanning a table prefixed by
+  // exactly key, all the batchIDs encountered will be unique and in order.
+  std::string mutations_prefix = LevelDbMutationKey::KeyPrefix(user_id_);
+  auto mutation_iterator = db_.currentTransaction->NewIterator();
+
+  std::vector<FSTMutationBatch*> result;
+  LevelDbDocumentMutationKey row_key;
+  for (; index_iterator->Valid(); index_iterator->Next()) {
+    // Only consider rows matching exactly the specific key of interest. Index
+    // rows have this form (with markers in brackets):
+    //
+    // <User>user <Path>collection <Path>doc <BatchId>2 <Terminator>
+    // <User>user <Path>collection <Path>doc <BatchId>3 <Terminator>
+    // <User>user <Path>collection <Path>doc <Path>sub <Path>doc <BatchId>3
+    // <Terminator>
+    //
+    // Note that Path markers sort after BatchId markers so this means that when
+    // searching for collection/doc, all the entries for it will be contiguous
+    // in the table, allowing a break after any mismatch.
+    if (!absl::StartsWith(index_iterator->key(), index_prefix) ||
+        !row_key.Decode(index_iterator->key()) ||
+        row_key.document_key() != key) {
+      break;
+    }
+
+    // Each row is a unique combination of key and batchID, so this foreign key
```

Comment:
Is this block approx AllMutationBatchesWithIds()? If so, I think you could refactor, though see above comment first.

Reply:
Refactored as above.